### PR TITLE
Resume search (#43)

### PR DESCRIPTION
**Pull request by alexrenz**
This PR adds functionality to resume a previously-run search, as discussed in #43 and #34. It adds a new command `resume-search` that re-runs the last-run command. It leverages fzf's history functionality: fzf records its last query to a history file, which we can then feed into the next fzf invocation via `--query`. It does not have a default key binding. 

This is my first time working on a VSCode extension, so I am looking forward to any feedback and comments! 

### Resume search in action
[](https://gifyu.com/image/S47Kr)

#### Resume position
For `find_within_files`, a resumed search actually pre-selects the item that was last selected in the previous search, see the above demo (using two `--bind`s in fzf and another file to record the position). I did not get this to work for `find_files` because the `find_files` script does not reload the rg results (in contrast to `find_within_files`). 

The problem is that the`{n}` in fzf's `--bind` gives the position of the selected item with respect to all candidates that fzf is *currently considering*, not the ones that it is *currently displaying*. Example: Say fzf currently has 50 candidates, and my query narrows this down to 5. Now I accept the third one. With reload (as in `find_within_files`), the `{n}` in fzf's `--bind` gives me position 3, which is what I want. Without the reload, `{n}` gives me the position of the selected item in the list of 50 candidates, for example 39, which is not what I want. 

Is it on purpose that `find_files` does not reload the results? Or could we potentially change this so that it works similarly to `find_within_files`? Or do you have other ideas? (because resuming with the previous position is really much much nicer than without)

#### Type filters
The type filter commands fail on my machine even with the unmodified extension from the VSCode marketplace, so I wasn't able to test the resume search functionality on the type filter commands. The problem: The terminal opens for me, but then I get a "Command failed" error in that terminal. Unfortunately, I cannot see the entire command. Do the type filters work for others or is this a known issue? 

#### ~Windows~
~I tried to implement the windows versions of this, but was not able to test them because I don't have access to a windows machine. And I don't know any PowerShell :D It would be amazing if someone could test this and report back what they find.~

### CODE REVIEW

**Review by tomrijndorp — COMMENTED** [review]
Nice work! Apologies though, it's going to take me a little longer to approve this because I can't run the extension on your brach, but I suspect it has to do with my VS Code installation.

As for your comment on keeping the position in the list and how the two commands are different: this was indeed deliberate. I'm a little fuzzy on all the details since it's been a few years, but I recall the reason was performance.

Let me follow up when I manage to get your branch to run on my machine. In the meantime I left a few comments.

File: find_files.ps1 [hunk 1]
```diff
@@ -62,8 +62,11 @@ $PREVIEW_ENABLED=VGet "env:FIND_WITHIN_FILES_PREVIEW_ENABLED" 0
 $PREVIEW_COMMAND=VGet "env:FIND_WITHIN_FILES_PREVIEW_COMMAND"  'bat --decorations=always --color=always --plain {}'
 $PREVIEW_WINDOW=VGet "env:FIND_WITHIN_FILES_PREVIEW_WINDOW_CONFIG" 'right:50%:border-left'
 $HAS_SELECTION=VGet "env:HAS_SELECTION" 0
+$RESUME_SEARCH=VGet "env:RESUME_SEARCH" 0
```
[the remark this answers is not in the document]
Comment (alexrenz) [reply find_files.ps1:65]: Fine for me! I don't feel confident in this working on Windows without testing, so I will switch to showing a not implemented message.

**Comment by tomrijndorp** [comment]
Apologies for the delay! I gave this a spin on both my Mac and Linux machines, and here's what I found:

Linux:
- resume find files works great!
- search within files was broken for me; throws `unsupported key: load`. This was added in FZF 0.36, which was released in January of this year. That makes it relatively "new", unfortunately newer than what you get by default on many distros (I'm on Ubuntu 22.04 LTS) when installing it. I think this is a deal breaker if `load` is required to implement this functionality, unless there's an elegant way to only pass `load` in when FZF actually supports it, and show a message notifying that FZF needs upgrading when invoking resume search. But I understand that's a bit of a pain. I do already grab the FZF version number components for other compatibility reasons, check `FZF_VER` in `shared.sh`. I did install a more recent FZF, and subsequently this was working great, too :)

Mac:
- same result, Homebrew gave me an older version as well when I last updated.

Please let me know your thoughts. This extension is already a little annoying in the sense that it requires external dependencies, and requiring relatively recent versions makes the out-of-the-box experience worse. If you don't have more cycles I can try giving that switching logic a try, but it won't be before the end of next week.

**Comment by alexrenz** [comment]
Ah, nice catch! I changed this a little so that "load" is used only when fzf>=0.36. It is a little hacky with a dummay fallback binding `start:ignore` because I didn't get it to work to put the entire binding (with the `--bind` part) conditionally.

**Comment by tomrijndorp** [comment]
> Ah, nice catch! I changed this a little so that "load" is used only when fzf>=0.36. It is a little hacky with a dummay fallback binding `start:ignore` because I didn't get it to work to put the entire binding (with the `--bind` part) conditionally.
> 
> For fzf<0.36, the search will be resumed without the previously selected item being re-selected. I haven't put a message about this yet though. Not sure where to best put that.

I think that's perfect actually! No need for that; I meant to say "if you remove the ability to resume entirely for lower versions _then_ there should be a message", but I didn't articulate it well. This is the better solution.

File: find_within_files.sh [hunk 2]
```diff
@@ -63,6 +76,12 @@ if [[ "$PREVIEW_ENABLED" -eq 1 ]]; then
     PREVIEW_STR=(--preview "$PREVIEW_COMMAND" --preview-window "$PREVIEW_WINDOW")
 fi
 
+RESUME_POS_BINDING="start:ignore" # dummy fallback binding because fzf v<0.36 does not support `load` and I did not figure out how to conditionally set the entire binding string (i.e., with the "--bind" part)
```
Comment (tomrijndorp) [inline find_within_files.sh:79]: Yeah, this is the way. Bash is a terrible, terrible language for stuff like this, and I really feel like I've been stretching what's within reason to do with bash while working on this extension... It's been painful :')

Comment (tomrijndorp) [reply find_within_files.sh:79]: I tried this and I actually get `unsupported key: start` now :) I had luck changing it to `backward-eof:ignore`. Definitely a hack but it seems to work :) I'm going to merge this and add this in.

**Comment by tomrijndorp** [comment]
Thanks so much for doing this and bearing with me through the compatibility fixes! I'm going to push a new release that will have these changes in the next hour or so :)

**Comment by alexrenz** [comment]
Awesome 🎉 Thanks for seeing this through so quickly! 

I agree this is pushing Bash, but on the other hand it is super inspiring to see how much value one can add by cleverly stitching things together with "simple" tools 🙂

**Pull request merged**